### PR TITLE
[feature](function) support ip functions named ipv4_to_ipv6 and cut_ipv6

### PR DESCRIPTION
**Pull request by sjyango**
## Proposed changes

Issue Number: close #xxx

- `ipv4_to_ipv6`: accept an IPv4 type address and return the converted IPv6 type address.
- `cut_ipv6`: accept an IPv6 type address and return a string containing the address of the specified number of bytes removed in text format.
- improve ut framework to support parsing of IPv4 and IPv6 types.
- add some uts and regression tests for these two functions.

### CODE REVIEW

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run p0

File: be/src/vec/functions/function_ip.h [hunk 1]
```diff
@@ -1159,4 +1162,136 @@ class FunctionToIP : public IFunction {
...
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return std::make_shared<DataTypeIPv6>();
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
+        const auto& ipv4_column_with_type_and_name = block.get_by_position(arguments[0]);
+        WhichDataType addr_type(ipv4_column_with_type_and_name.type);
+        const auto* ipv4_column =
+                assert_cast<const ColumnIPv4*>(ipv4_column_with_type_and_name.column.get());
```
Comment (zhiqiang-hhhh) [inline be/src/vec/functions/function_ip.h:1183]: what if column const

Comment (sjyango) [reply be/src/vec/functions/function_ip.h:1183]: Good instructions, I will improve the handling of const

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java [hunk 2]
```diff
@@ -663,6 +665,8 @@ public class BuiltinScalarFunctions implements FunctionHelper {
             scalar(ToIpv6.class, "to_ipv6"),
             scalar(ToIpv6OrDefault.class, "to_ipv6_or_default"),
             scalar(ToIpv6OrNull.class, "to_ipv6_or_null"),
+            scalar(Ipv4ToIpv6.class, "ipv4_to_ipv6"),
+            scalar(CutIpv6.class, "cut_ipv6"),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:669]: other function sort by lexicographical order, could u help reorder ip functions with lexicographical order too?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * scalar function cut_ipv6
+ */
+public class CutIpv6 extends ScalarFunction
+        implements BinaryExpression, ExplicitlyCastableSignature, PropagateNullable {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java:40]: it is not an BinaryExpression, should impl TernaryExpression

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    public static final List<FunctionSignature> SIGNATURES = ImmutableList.of(
+            FunctionSignature.ret(VarcharType.SYSTEM_DEFAULT)
+                .args(IPv6Type.INSTANCE, TinyIntType.INSTANCE, TinyIntType.INSTANCE),
+            FunctionSignature.ret(StringType.INSTANCE)
+                .args(IPv6Type.INSTANCE, TinyIntType.INSTANCE, TinyIntType.INSTANCE));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java:46]: should not return diff type with same arg types, remove one of them

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Ipv4ToIpv6.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * scalar function ipv4_to_ipv6
+ */
+public class Ipv4ToIpv6 extends ScalarFunction
+        implements BinaryExpression, ExplicitlyCastableSignature, PropagateNullable {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Ipv4ToIpv6.java:38]: should be UnaryExpression

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java [hunk 6]
```diff
@@ -1309,6 +1311,14 @@ default R visitToIpv6OrNull(ToIpv6OrNull toIpv6OrNull, C context) {
         return visitScalarFunction(toIpv6OrNull, context);
     }
 
+    default R visitIpv4ToIpv6(Ipv4ToIpv6 ipv4ToIpv6, C context) {
+        return visitScalarFunction(ipv4ToIpv6, context);
+    }
+
+    default R visitCutIpv6(CutIpv6 cutIpv6, C context) {
+        return visitScalarFunction(cutIpv6, context);
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/visitor/ScalarFunctionVisitor.java:1320]: same as BuildInScalarFunctions, sort by lexicographical order

File: gensrc/script/doris_builtins_functions.py [hunk 7]
```diff
@@ -2121,6 +2121,9 @@
         [['to_ipv6_or_default'], 'IPV6', ['STRING'], 'ALWAYS_NOT_NULLABLE'],
         [['to_ipv6_or_null'], 'IPV6', ['VARCHAR'], 'ALWAYS_NULLABLE'],
         [['to_ipv6_or_null'], 'IPV6', ['STRING'], 'ALWAYS_NULLABLE'],
+        [['ipv4_to_ipv6'], 'IPV6', ['IPV4'], 'DEPEND_ON_ARGUMENT'],
+        [['cut_ipv6'], 'VARCHAR', ['IPV6'], 'DEPEND_ON_ARGUMENT'],
+        [['cut_ipv6'], 'STRING', ['IPV6'], 'DEPEND_ON_ARGUMENT'],
```
Comment (morrySnow) [inline gensrc/script/doris_builtins_functions.py:2126]: this function signature is not same with nereids planner? should have three args?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java (hunk 2 above)
Comment (sjyango) [reply fe/fe-core/src/main/java/org/apache/doris/catalog/BuiltinScalarFunctions.java:669]: ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java (hunk 3 above)
Comment (sjyango) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java:40]: ok，I will change BinaryExpression to TernaryExpression

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java (hunk 4 above)
Comment (sjyango) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/CutIpv6.java:46]: ok

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Ipv4ToIpv6.java (hunk 5 above)
Comment (sjyango) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/Ipv4ToIpv6.java:38]: ok，I will change BinaryExpression to UnaryExpression

File: gensrc/script/doris_builtins_functions.py (hunk 7 above)
Comment (sjyango) [reply gensrc/script/doris_builtins_functions.py:2126]: yes, it's a mistakes, thanks for code review

**Comment by sjyango** [comment]
run buildall

**Review by zhiqiang-hhhh — APPROVED** [review]

File: be/src/vec/functions/function_ip.h [hunk 8]
```diff
@@ -1159,4 +1162,138 @@ class FunctionToIP : public IFunction {
...
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        return std::make_shared<DataTypeIPv6>();
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
+        const auto& ipv4_column_with_type_and_name = block.get_by_position(arguments[0]);
+        const auto& [ipv4_column, ipv4_const] =
+                unpack_if_const(ipv4_column_with_type_and_name.column);
+        const auto* ipv4_addr_column = assert_cast<const ColumnIPv4*>(ipv4_column.get());
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:1183]: if here is nullable column ? u can make some cases in groovy , insert values with null

Comment (sjyango) [reply be/src/vec/functions/function_ip.h:1183]: done

Comment (sjyango) [reply be/src/vec/functions/function_ip.h:1183]: nullable column is no problem

**Comment by sjyango** [comment]
run buildall

**Review by amorynan — APPROVED** [review]
LGTM

**Review by morrySnow — APPROVED** [review]

**Pull request merged**